### ISSUE TITLE
Various instances of incompatibly-typed variable reuse

### ISSUE THREAD

**Issue by alexrp** [Vector35/binaryninja-api#4488]
**Version and Platform (required):**
- Binary Ninja Version: 3.5.4388-dev Personal (a010e068)
- OS: Windows
- OS Version: 10.0.25375.1
- CPU Architecture: x64

**Bug Description:**
In these cases, `context` is reassigned to a completely different type, confusing subsequent analysis and xrefs:

![image](https://github.com/Vector35/binaryninja-api/assets/44076/d8c749b7-5f1f-4de7-b993-12401390bb24)
![image](https://github.com/Vector35/binaryninja-api/assets/44076/95e2736a-2859-4991-becd-44ad323f2d18)
![image](https://github.com/Vector35/binaryninja-api/assets/44076/1379404d-fd50-44ab-91b2-8c765fbf0970)
![image](https://github.com/Vector35/binaryninja-api/assets/44076/b5821773-5570-4a1d-94d9-1d3250f76cf2)
![image](https://github.com/Vector35/binaryninja-api/assets/44076/f9395169-dac2-44f1-8013-ff7364d67015)
![image](https://github.com/Vector35/binaryninja-api/assets/44076/b5507747-6fff-40e6-a55b-68041117aaad)

Issues like these can be found in hundreds of places where `S1Context::Singleton` is referenced, but it has ~23k references, so you may have better things to do than dig through all of those. 🙂

**Steps To Reproduce:**
1. Grab [this BNDB](https://drive.google.com/file/d/1mRX0JMkcleAZWVVl-DQcZICbllsVIibK/view?usp=sharing).
2. Navigate to the addresses seen in the screenshots.

**Comment by plafosse** [comment]
Are you unable to split those variable uses?

**Comment by alexrp** [comment]
OK, I'll be honest... I spend most of my time in the Pseudo C view, only switching to the ILs when needed, so I never even knew that variable splitting was a thing since it's only available in the ILs.

But, yes, I checked a few of the cases, and splitting seems to work.

Does this mean there's no bug here? Or should the decompiler ideally handle these cases?

**Comment by xusheng6** [comment]
I renamed the title to better reflect the current status of the issue

**Comment by alexrp** [comment]
Assuming the effort label on this one is accurate, is it likely that this could be addressed in the next stable? It's one of those small UX paper cuts that consistently wastes a bit of time whenever I use the feature.

Not a huge deal - just would be nice to get fixed if it *is* in fact low-hanging fruit. :eyes:

**Comment by psifertex** [comment]
We can't make any promises. Unfortunately there are [_lot_](https://github.com/Vector35/binaryninja-api/issues?q=is%3Aopen+is%3Aissue+label%3A%22Effort%3A+Low%22+label%3A%22Impact%3A+Medium%22) of "low effort" / "medium impact" issues and even low effort can be up to a week when including testing/review. With 116 of them, even if we had 5 people doing nothing but these it would still be almost six months! And that's not counting all the new ones that will get filed over the next few months.

TL;DR "maybe"

Make sure to upvote, we absolutely do [review](https://github.com/Vector35/binaryninja-api/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) the votes when deciding priority.

**Comment by alexrp** [comment]
> We can't make any promises.

Of course. I should have probably been clearer; I only meant whether it would be on the docket for the next stable.

> Unfortunately there are [_lot_](https://github.com/Vector35/binaryninja-api/issues?q=is%3Aopen+is%3Aissue+label%3A%22Effort%3A+Low%22+label%3A%22Impact%3A+Medium%22) of "low effort" / "medium impact" issues and even low effort can be up to a week when including testing/review. With 116 of them, even if we had 5 people doing nothing but these it would still be almost six months! And that's not counting all the new ones that will get filed over the next few months.

That's fair. I had assumed "low effort" was on the order of 1-2 days, hence the thinking that it's low-hanging fruit.

> Make sure to upvote, we absolutely do [review](https://github.com/Vector35/binaryninja-api/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc) the votes when deciding priority.

Just to clarify, *should* I actually be upvoting my own issues? :eyes: I've deliberately avoided doing so because I haven't noticed other people doing it with theirs, so it almost seemed like "bad manners" in a sense? But maybe I'm overthinking it? :shrug:
